### PR TITLE
Fallback fonts

### PR DESCRIPTION
**Pull request by KOLANICH**

### CODE REVIEW

**Comment by KOLANICH** [comment]
> Python 2.7 was released on July 3, 2010.

Shouldn't we drop 2.6?

**Comment by pietermarsman** [comment]
Python 2.6 is removed from the Travis spec on master, so you can do that to.

I like that we have some fallback metrics, but I wonder if this solves an actual problem or if it is just nice to have.

In any case, could you add a test that demonstrates how this works? If the test is sufficient and successful we can merge this.

**Comment by KOLANICH** [comment]
It solves a real problem. I have some crippled PDFs. I mean I have downloaded some PDFs and have removed all the fonts blobs from them to reduce their sizes. Semi-automated. Ijust decompressed all the pdfs and remove the tags with a bash script and then recompressed cleaning up and fixing errors. It is unlikely one will see such PDFs in the wild now when every producer embeds fonts. These documents still look fine in a viewer (but the fonts are different, but is not an issue for me) but consume a few mibs less **each**. But pdfminer.six had failed on grouping such characters or lines - I don't remember what was the exact issue.

About test - currently I have no time to work on this.

File: pdfminer/utils.py [hunk 1]
```diff
@@ -293,6 +293,21 @@ def matrix2str(m):
...
+
+def mean(it):
+    sum=0
+    count=0
+    for el in it:
+        sum+=el
+        count+=1
+    return sum / count
+
+def itemz(it):
```
Comment (pietermarsman) [inline pdfminer/utils.py]: I find the name of this function perculiar. Maybe rename it to `key_values()`?

File: pdfminer/utils.py [hunk 2]
```diff
@@ -293,6 +293,21 @@ def matrix2str(m):
...
 
 
+
+def mean(it):
+    sum=0
+    count=0
+    for el in it:
+        sum+=el
+        count+=1
+    return sum / count
```
Comment (pietermarsman) [inline pdfminer/utils.py]: You could compute this with a readable oneliner:

```python
return sum(it) / len(list(it))
```

File: pdfminer/pdffont.py [hunk 3]
```diff
@@ -76,7 +78,33 @@ def get_widths2(seq):
 #assert get_widths2([1,2,3,4,5]) == {1:(3, (4,5)), 2:(3, (4,5))}
 #assert get_widths2([1,[2,3,4,5],6,[7,8,9]]) == {1:(2, (3,4)), 6:(7, (8,9))}
 
-
+def computeStat(it):
```
Comment (pietermarsman) [inline pdfminer/pdffont.py]: Function names should be lowercase with underscores for extra readability: https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: pdfminer/pdffont.py [hunk 4]
```diff
@@ -76,7 +78,33 @@ def get_widths2(seq):
...
+			metrics[cat]=res[0]
+	return metrics
+
+def computeFallbackFontMetrics(FONT_METRICS):
+    #TODO:
+    # * weight-dependent metrics
+    # * machine-learning based metrics prediction
+    #weights=defaultdict(list)
+    #for f in FONT_METRICS.values():
+    #    weights[f[0]['FontWeight']].append(f)
```
Comment (pietermarsman) [inline pdfminer/pdffont.py]: Remove or include those comments.

File: pdfminer/pdffont.py [hunk 5]
```diff
@@ -76,7 +78,33 @@ def get_widths2(seq):
...
+	metrics={}
+	for cat, sizes in stat.items():
+		res=max(sizes.items(), key=lambda x: x[1])
+		if res[1] == 1:
+			metrics[cat]=mean(sizes.keys())
+		else:
+			metrics[cat]=res[0]
+	return metrics
+
+def computeFallbackFontMetrics(FONT_METRICS):
```
Comment (pietermarsman) [inline pdfminer/pdffont.py]: This function is rather complicated. Could you add a docstring that describe what it does?

File: pdfminer/pdffont.py [hunk 6]
```diff
@@ -76,7 +78,33 @@ def get_widths2(seq):
...
+    #TODO:
+    # * weight-dependent metrics
+    # * machine-learning based metrics prediction
+    #weights=defaultdict(list)
+    #for f in FONT_METRICS.values():
+    #    weights[f[0]['FontWeight']].append(f)
+    charMetrics=computeStat((fd[1] for fd in FONT_METRICS.values()))
+    restMetrics=computeStat(({k:v for k,v in fd[0].items() if isinstance(v, (int, float))} for fd in FONT_METRICS.values()))
```
Comment (pietermarsman) [inline pdfminer/pdffont.py]: If I understand correctly, this filters out the values font metrics that are not `int` or `float`. Why is that? 

This skips FontName, which should be in the font description according to pdf reference. Also see Table 5.19 from the [pdf reference](https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf).

File: pdfminer/pdffont.py [hunk 7]
```diff
@@ -76,7 +78,33 @@ def get_widths2(seq):
 #assert get_widths2([1,2,3,4,5]) == {1:(3, (4,5)), 2:(3, (4,5))}
 #assert get_widths2([1,[2,3,4,5],6,[7,8,9]]) == {1:(2, (3,4)), 6:(7, (8,9))}
 
-
+def computeStat(it):
+	stat=defaultdict(lambda: defaultdict(int))
+	for fd in it:
+		for cat, size in itemz(fd):
```
Comment (pietermarsman) [inline pdfminer/pdffont.py]: These variable names (`cat` and `size`) confuse me. They are the key-value pairs for each font metric for a specific font, right? 

Maybe call them `font_descriptor_name` and `font_descriptor_name`?

File: pdfminer/utils.py (hunk 2 above)
Comment (KOLANICH) [reply pdfminer/utils.py]: This function can work with finite length iterators. Though it turned out that isn't needed.

File: pdfminer/utils.py (hunk 1 above)
Comment (KOLANICH) [reply pdfminer/utils.py]: or maybe `universal_keys`

File: pdfminer/pdffont.py (hunk 6 above)
Comment (KOLANICH) [reply pdfminer/pdffont.py]: It computes mean. For non-numerical we can compute the most frequent, but it is not needed, we currently need only sizes in order to glue the chars into words and lines correctly.

File: pdfminer/pdffont.py (hunk 4 above)
Comment (KOLANICH) [reply pdfminer/pdffont.py]: It just turned out they are have no use tor us currently.

File: pdfminer/pdffont.py (hunk 5 above)
Comment (KOLANICH) [reply pdfminer/pdffont.py]: Unfortunately I don't remember why I have implemented this this way. So I can only to try to RE it from the code.

**Comment by pietermarsman** [comment]
Hi @KOLANICH,

Your PR contains a lot of work that is already merged to develop. Could you clean up your PR? 
Hint: force-pushing is a practice you want to avoid. 

More importantly, to respond to your problem: 

> It solves a real problem. I have some crippled PDFs. I mean I have downloaded some PDFs and have removed all the fonts blobs from them to reduce their sizes. 

I don't think this project should focus on parsing pdf's that were deliberately broken.

**Comment by KOLANICH** [comment]
https://github.com/pdfminer/pdfminer.six/pull/272

**Pull request closed without merging**